src/ngraph/ngraph_graph.cc

```diff
@@ -262,96 +269,148 @@ std::vector<NodePtr> FindSubgraph(const Graph& graph, NodePtr node,
-      tmpGraph->shape_ = output->shape_;
-      tmpGraph->dtype_ = output->dtype_;
-      tmpGraph->stype_ = output->stype_;
-
-      auto in_tmpGraphInputs = [&tmpGraph](NodePtr n) {
-        if (!in_vec(tmpGraph->inputs_, n)) return false;
-        return true;
-      };
-      // setup inputs to this subgraph (as a node)
-      for (auto node : tmpGraph->nodes_) {
-        for (auto input : node->inputs_) {
-          if (input->subgraph_ != i && !in_tmpGraphInputs(input))
-            tmpGraph->inputs_.emplace_back(input);
+  auto tmpGraph = std::make_shared<Graph>(
+      "subgraph_" + randomString(12) + std::to_string(subgraph_num),
+      graph->context_);
+  // loop over all nodes and add nodes in the current subgraph to
+  for (auto node : graph->nodes_) {
+    if (node->subgraph_ == subgraph_num) {
+      tmpGraph->AddNode(node);
+    }
+  }
+
+  if (tmpGraph->nodes_.size() != 0) {
+    tmpGraph->outputs_ = GetSubgraphOutputs(*graph, tmpGraph->nodes_);
+    tmpGraph->num_outputs_ = tmpGraph->outputs_.size();
+    for (size_t i = 0; i < tmpGraph->outputs_.size(); ++i) {
+      tmpGraph->output_elements_.emplace_back(
+          std::make_shared<OutputElement>(tmpGraph, i));
+      tmpGraph->output_elements_.back()->subgraph_ = subgraph_num;
+    }
+    // if we found nodes, setup subgraph
+    tmpGraph->in_ngraph_ = true;
+    tmpGraph->subgraph_ = subgraph_num;
+
+    auto in_tmpGraphInputs = [&tmpGraph](NodePtr n) {
+      if (!in_vec(tmpGraph->inputs_, n)) return false;
+      return true;
+    };
+
+    // setup inputs to this subgraph (as a node)
+    for (auto node : tmpGraph->nodes_) {
+      for (auto input : node->inputs_) {
+        if (input->subgraph_ != subgraph_num && !in_tmpGraphInputs(input))
+          tmpGraph->inputs_.emplace_back(input);
+      }
+    }
+    // create a map between base nodes and outputs for easier repalcement
+    std::unordered_map<NodePtr, NodePtr> output_map;
+    for (auto output : tmpGraph->output_elements_) {
+      output_map.insert({output->base_node_, output});
+    }
+
+    // set any needed graph outputs to sub_graph outputs
+    for (size_t i = 0; i < graph->outputs_.size(); ++i) {
+      if (output_map.count(graph->outputs_[i])) {
+        graph->outputs_[i] = output_map[graph->outputs_[i]];
+      }
+    }
+    // insert the new outputs
+    for (auto output : tmpGraph->output_elements_) {
+      auto it = std::find_if(
+          graph->nodes_.begin(), graph->nodes_.end(),
+          [output](NodePtr n) -> bool { return (n == output->base_node_); });
+      graph->nodes_.insert(it, output);
+    }
+    // delete all the nodes we're replacing with the subgraph
+    graph->nodes_.erase(
+        std::remove_if(graph->nodes_.begin(), graph->nodes_.end(),
+                       [](NodePtr n) -> bool {
+                         return ((n->subgraph_ > 0) &&
+                                 (n->type_ == NodeType::kOp));
+                       }),
+        graph->nodes_.end());
```

Comment:
Did you really mean to use both `erase(...)` _and_ `remove_if(...)`?  Perhaps it's correct, but it smells fishy.

Reply:
https://en.wikipedia.org/wiki/Erase%E2%80%93remove_idiom

Reply:
Thanks, I didn't know about that idiomatic use of the garbage fire that is the STL :)